File: app/src/main/java/igrek/songbook/persistence/preferences/PreferencesDefinition.java
package igrek.songbook.persistence.preferences;

public enum PreferencesDefinition {
	
	fontsize(20.0f), // dp
	
	autoscrollInitialPause(36000L), // ms
	
	autoscrollSpeed(0.15f); // em / s
	
	private Object defaultValue;
	private PropertyType type;
	
	PreferencesDefinition(String defaultValue) {
		this(PropertyType.STRING, defaultValue);
	}
	
	PreferencesDefinition(Boolean defaultValue) {
		this(PropertyType.BOOLEAN, defaultValue);
	}
	
	PreferencesDefinition(Long defaultValue) {
		this(PropertyType.LONG, defaultValue);
	}
	
	PreferencesDefinition(Float defaultValue) {
		this(PropertyType.FLOAT, defaultValue);
	}
	
	PreferencesDefinition(PropertyType type, Object defaultValue) {
		this.type = type;
		this.defaultValue = defaultValue;
	}
	
	public Object getDefaultValue() {
		return defaultValue;
	}
	
	public PropertyType getType() {
		return type;
	}
	
	
}

File: app/src/main/java/igrek/songbook/settings/preferences/PreferencesDefinition.kt
package igrek.songbook.settings.preferences

import igrek.songbook.settings.chordsnotation.ChordsNotation
import igrek.songbook.settings.theme.ColorScheme
import igrek.songbook.settings.theme.FontTypeface

enum class PreferencesDefinition constructor(val type: PropertyType, val defaultValue: Any?) {

    Fontsize(20.0f), // dp

    AutoscrollInitialPause(36000L), // ms

    AutoscrollSpeed(0.15f), // em / s

    ChordsNotationId(ChordsNotation.GERMAN.id),

    AppLanguage(igrek.songbook.settings.language.AppLanguage.DEFAULT.langCode),

    FontTypefaceId(FontTypeface.SANS_SERIF.id),

    ColorSchemeId(ColorScheme.DARK.id),

    ChordsEndOfLine(false),

    AutoscrollSpeedAutoAdjustment(true),

    AutoscrollSpeedVolumeKeys(true),

    RandomFavouriteSongsOnly(false),

    CustomSongsGroupCategories(false),

    RestoreTransposition(true),

    ;

    constructor(defaultValue: String) : this(PropertyType.STRING, defaultValue)

    constructor(defaultValue: Boolean?) : this(PropertyType.BOOLEAN, defaultValue)

    constructor(defaultValue: Long?) : this(PropertyType.LONG, defaultValue)

    constructor(defaultValue: Float?) : this(PropertyType.FLOAT, defaultValue)

    fun preferenceName(): String {
        return this.name.decapitalize()
    }

}
